### ISSUE TITLE
Jedis should not require localhost to be resolved so it can run clean in serverless env

### ISSUE THREAD

**Issue by xoraes** [xetorthio/jedis#1424]
### Expected behavior
Jedis should not require localhost to be resolved so it can run clean in serverless env - aws lambda. 

### Actual behavior
When running jedis on aws lambda, it tries to resolve localhost which is not available in serverless env resulting in a "SEVERE" log output.

Write here what happens instead ...

    15:02:45 SEVERE: cant resolve localhost address
    15:02:45 java.net.UnknownHostException: ip-10-29-121-116: ip-10-29-121-116: unknown error
    15:02:45 at java.net.InetAddress.getLocalHost(InetAddress.java:1505)
    15:02:45 at redis.clients.jedis.HostAndPort.getLocalHostQuietly(HostAndPort.java:105)

### Steps to reproduce:
Use jedis client in aws lambda to connect to aws redis service endpoint
    public class RedisClientDao {

    private static JedisPool pool = null;

    static {
        String endpoint = Preconditions.checkNotNull((String)Config.get("CACHE_ENDPOINT"));
        JedisPoolConfig poolConfig = new JedisPoolConfig();
        poolConfig.setJmxEnabled(false);
        pool = new JedisPool(poolConfig, endpoint);

    }

    public static Jedis getRedisClient() {
        return pool.getResource();
    }
    }

### Redis / Jedis Configuration
jedis 2.9.0 running on java 1.8 running on aws lambda connecting to redis 3.2.4

#### Jedis version:
2.9
#### Redis version:
3.2.4
#### Java version:
1.8

**Comment by smagellan** [comment]
I suppose it is related with https://github.com/xetorthio/jedis/issues/1367, https://github.com/xetorthio/jedis/pull/1099, https://github.com/xetorthio/jedis/pull/1076 and was merged as part of https://github.com/xetorthio/jedis/pull/1342. My proposal is to:

- Introduce HostAndPortBuilder. It should resolve localhost if configured to do so.

or
- Introduce extra more conditional logic to HostAndPort.parseString, e.g.:
```Java
    boolean doLocalhostAliasing = System.getProperty("redis.clients.jedis.doLocalhostAliasing") != null;
    return new HostAndPort( doLocalhostAliasing ? convertHost(host) : host, port);
```
@marcosnils, @xetorthio, @allanwax any ideas?

**Comment by allanwax** [comment]
The original intent of the fix I submitted was to resolve the problem of specifying localhost in a cluster.  The fix changed localhost to the ip of the localhost so that node information passed around would succeed.  Putting a conditional on this is fine with me but may affect testing.

What will the default value of this flag be?

**Comment by xoraes** [comment]
is there a prior release/version I can use which may not have this issue? This issue is a blocker for me.

**Comment by marcosnils** [comment]
@xoraes can you try 2.8.x? That release shouldn't have this logic. I'm on vacation atm, I'll check this issue next week

@heartsavior can u take a look if u have some time plz??

**Comment by jkrahul** [comment]
@marcosnils encountered the same issue. working fine with 2.8.1

**Comment by marcosnils** [comment]
@jkrahul thx. I'll close this if confirmed it works with 2.8.1

**Comment by marcosnils** [comment]
Whoops, we need to fix this somehow so it works in future versions. Anyone willing to craft a PR?. I don't have too much time in my hands lately.

**Comment by jkrahul** [comment]
This error is occurring not just in serverless env (AWS Lambda), but in AWS ec2 instances also.

**Comment by marcosnils** [comment]
@jkrahul thx for reporting. Any chances you'd like to craft a patch for this?

**Comment by jkrahul** [comment]
@marcosnils Will do

**Comment by allanwax** [comment]
@marcosnils I have looked at the way `AWS` uses `REDIS` and the few examples that I saw do not use an IP address to set up the connection but instead use an `ENDPOINT` address.  I'd also like to say that I'm not an AWS user.

The examples I saw treat REDIS as a single instance and setting up the Jedis instance looks something like
```
try (Jedis jedis = new Jedis(Global.CLUSTER_ENDPOINT, 6379, 10000);) {              
    jedis.hset(f.dayOfFile, "content", f.xmlFile);
}
```

`HostAndPort` would not seem to be used at all in creating a Jedis instance.  Again, this is based upon the limited amount of example code I could find.  So using `HostAndPort` in an AWS context may be meaningless.  However that is not to say that it couldn't be.

So along that line of reasoning I've come up with a proposed change to HostAndPort that may make the problem better but not necessarily make the problem go away.

```
public static String convertHost(String host) {
	try {
		/*
		 * Validate the host name as an IPV4/IPV6 address.
		 * If this is an AWS ENDPOINT it will not parse.
		 * In that case accept host as is.
		 * 
		 * Costs: If this is an IPV4/6 encoding, e.g. 127.0.0.1 then no DNS lookup
		 * is done.  If it is a name then a DNS lookup is done but it is normally cached.
		 * Secondarily, this class is typically used to create a connection once
		 * at the beginning of processing and then not used again.  So even if the DNS
		 * lookup needs to be done then the cost is miniscule.
		 */
		InetAddress inetAddress = InetAddress.getByName(host);

		// isLoopbackAddress() handles both IPV4 and IPV6
		if (inetAddress.isLoopbackAddress() || host.equals("0.0.0.0") || host.startsWith("169.254"))
			return LOCALHOST_STR;
		else
			return host;
	}
	catch (Exception e) {
		// Not a valid IP address
		log.logp(Level.WARNING, HostAndPort.class.getName(), "convertHost", "'" + host + "' is not a valid IP address", e);
		return host;
	}
}
```
 
This basically 'punts' on `ENDPOINT` addresses and handles IP addresses in the normal way.  It does not change the value of `LOCALHOST_STR` but perhaps a method should be added: `setLocalhost(String localhost)`

**Comment by jkrahul** [comment]
@marcosnils @allanwax Along with this, Lazy resolution of static variable LOCALHOST_STR should be done to avoid the exception in the first place.

**Comment by allanwax** [comment]
Something like this maybe

```
...
protected static String localhost = null;
...
// isLoopbackAddress() handles both IPV4 and IPV6
		if (inetAddress.isLoopbackAddress() || host.equals("0.0.0.0") || host.startsWith("169.254"))
			return getLocalhost();
		else
			return host;
...
public static void setLocalhost(String localhost) {
    this.localhost = localhost;
}

public String getLocalhost() {
    if (localhost != null) {
        return localhost;
    }
    else {
        return localhost = getLocalHostQuietly();
    }
}
...
```

**Comment by marcosnils** [comment]
@allanwax @jkrahul thx the feedback. I don't have time to craft a PR currently. I hope @jkrahul has some time to do it soon.

**Comment by jkrahul** [comment]
TEST is failling in master branch
`Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.005 sec <<< FAILURE! - in redis.clients.jedis.tests.ModuleTest
testModules(redis.clients.jedis.tests.ModuleTest)  Time elapsed: 0.005 sec  <<< ERROR!
redis.clients.jedis.exceptions.JedisDataException: ERR unknown command 'MODULE'
	at redis.clients.jedis.tests.ModuleTest.testModules(ModuleTest.java:34)`

**Comment by jkrahul** [comment]
@marcosnils @allanwax TEST case is failing
`testModules(redis.clients.jedis.tests.ModuleTest)  Time elapsed: 0.006 sec  <<< ERROR!
redis.clients.jedis.exceptions.JedisDataException: ERR unknown command 'MODULE'
	at redis.clients.jedis.tests.ModuleTest.testModules(ModuleTest.java:34)
`

**Comment by marcosnils** [comment]
@jkrahul you need to use latest redis from `unstable` branch in order for it to work.  https://github.com/xetorthio/jedis/blob/1f1e23b26a4aa244ebfdbcbcdf489b630c841884/.github/CONTRIBUTING.md#jedis-test-environment

**Comment by user20161119** [comment]
any updates?

**Comment by allanwax** [comment]
@user20161119 I don't know if this has been included yet but when it is, you can manually set the value of localhost to whatever you want and then it will not be looked up

**Comment by user20161119** [comment]
thanks @allanwax  we integrate jedis with spring data,i am looking for the setting in xml config as follow:

```
	<bean id="jedisConnectionFactory"
		class="org.springframework.data.redis.connection.jedis.JedisConnectionFactory">
		<property name="hostName" value="${redis_baseinfo_ip}" />
		<property name="port" value="${redis_baseinfo_port}" />
		<property name="password" value="${redis_baseinfo_password}" />
		<property name="poolConfig">
			<bean class="redis.clients.jedis.JedisPoolConfig">
				<property name="maxTotal" value="${redis_baseinfo_pool_maxActive}" />
				<property name="maxIdle" value="${redis_baseinfo_pool_maxIdle}" />
				<property name="testOnBorrow" value="true" />
			</bean>
		</property>
	</bean>
```

**Comment by allanwax** [comment]
I don't use `XML` to configure my Redis but to do so I imagine you will have to create a bean for `HostAndPort` and set a property for `localHost`.  I believe it is static so once set it should get used.

And maybe this bean should be created before the pool.  Don't know

**Comment by jhackel-hypo** [comment]
Are there any plans to release a bug fix release for the 2.9 branch? This version is part of Spring Boot and makes my Docker application log an error every time the application starts, triggering false alerts.
